tests/tests/filter.cpp

```diff
@@ -0,0 +1,446 @@
+  static_assert(not err_handler_for<decltype(wrongOnError), operand_t>);
+  static_assert(not err_handler_for<decltype(wrongOnError2), operand_t>);
+}
+
+TEST_CASE("filter", "[concepts][expected_void]")
+{
+  using namespace fn::detail;
+  using operand_t = std::expected<void, Error>;
+
+  auto predicate = [] { return true; };
+  auto wrongPredicate = []() -> int { return 1; };
+  auto onError = [] { return Error{"Some error"}; };
+  auto onError2 = [] { return DerivedError{"Some error"}; };
+  auto wrongOnError = [] { return IncompatibleError{}; };
+
+  static_assert(void_predicate_for<decltype(predicate), operand_t>);
+  static_assert(not void_predicate_for<decltype(wrongPredicate), operand_t>);
+  static_assert(void_err_handler_for<decltype(onError), operand_t>);
+  static_assert(void_err_handler_for<decltype(onError2), operand_t>);
+  static_assert(not void_err_handler_for<decltype(wrongOnError), operand_t>);
+}
+
+TEST_CASE("filter", "[concepts][optional]")
+{
+  using namespace fn::detail;
+  using operand_t = std::optional<int>;
+
+  auto predicate = [](int) { return true; };
+  auto wrongPredicate = [](std::string) { return true; };
+  auto wrongPredicate2 = [](int) -> int { return 1; };
+
+  static_assert(predicate_for<decltype(predicate), operand_t>);
+  static_assert(not predicate_for<decltype(wrongPredicate), operand_t>);
+  static_assert(not predicate_for<decltype(wrongPredicate2), operand_t>);
+}
+
+TEST_CASE("filter", "[filter][expected][expected_value]")
+{
+  using namespace fn;
+
+  using operand_t = std::expected<int, Error>;
+  constexpr auto truePred = [](int) { return true; };
+  constexpr auto falsePred = [](int) { return false; };
+  constexpr auto onError
+      = [](int v) { return Error{"Got " + std::to_string(v)}; };
+
+#define WITH_BOUND_E(op, f)                                                    \
+  [](auto &&fn) constexpr {                                                    \
+    return monadic_invocable<filter_t, op, decltype(fn), decltype(onError)>;   \
+  }(f)
+
+#define WITH_BOUND_P(op, f)                                                    \
+  [](auto &&fn) constexpr {                                                    \
+    return monadic_invocable<filter_t, op, decltype(truePred), decltype(fn)>;  \
+  }(f)
+
+#define CHECK_CALL_P(op, f) static_assert(WITH_BOUND_E(op, f))
+#define CHECK_CALL_E(op, f) static_assert(WITH_BOUND_P(op, f))
+#define CHECK_CANT_CALL_P(op, f) static_assert(not WITH_BOUND_E(op, f))
+#define CHECK_CANT_CALL_E(op, f) static_assert(not WITH_BOUND_P(op, f))
+
+  // lvalue operand
+  // --------------
+  static_assert(monadic_invocable<filter_t, operand_t, decltype(truePred),
+                                  decltype(onError)>);
+
+  static_assert(monadic_invocable<filter_t, operand_t, decltype(falsePred),
+                                  decltype(onError)>);
+
+  CHECK_CALL_P(operand_t, [](auto...) -> bool { throw 0; });
```

Comment:
@Bronek as much as i dislike macros... this is more readable and maybe we should do something like this everywhere, wdyt?

Reply:
I would keep it. Just remember to add `#undef` below so we can define them differently in tests that follow. Might want to create a `tests/util.hpp` for something similar in the future (feel free to raise an issue) and rewrite existing tests.

Reply:
I have for now created a util for this which got two benefits:
- no macros
- static_assert, while annoying to type, it now shows up on the right line in the IDE when it fails

While the first is an obvious win, the second i still have doubts about. Maybe the assert should be hidden inside the util for slightly improved readability and less typing but with the downside of IDE not showing where the assertion actually happens.